mobius-core/src/test/java/com/spotify/mobius/MobiusLoopTest.java

```diff
@@ -374,6 +399,163 @@ public void dispose() {
     observer.assertStates("Firstinit->1");
   }
 
+  @Test
+  public void eventSourceMayEmitEventsBeforeReturningFromDispose() throws Exception {
+    // Event sources should be allowed to emit events before they return
+    // from their dispose() method.
+
+    final MobiusLoop.Builder<String, TestEvent, TestEffect> builder =
+        Mobius.loop(update, effectHandler)
+            .eventSource(new EmitDuringDisposeEventSource(new TestEvent("foo")));
+
+    builder.startFrom("foo").dispose();
+  }
+
+  @Test
+  public void effectHandlerMayEmitEventsBeforeReturningFromDispose() throws Exception {
+    // Effect handlers should be allowed to emit events before they return
+    // from their dispose() method.
+
+    final MobiusLoop.Builder<String, TestEvent, TestEffect> builder =
+        Mobius.loop(update, new EmitDuringDisposeEffectHandler());
+
+    builder.startFrom("foo").dispose();
+  }
+
+  @Test
+  public void updateMayEmitModelsBeforeReturningFromDispose() throws Exception {
+    // Update should be allowed to emit model changes during dispose without anything
+    // bad happening.
+
+    Semaphore lock = new Semaphore(0);
+
+    final MobiusLoop.Builder<String, TestEvent, TestEffect> builder =
+        Mobius.loop(
+            (model, event) -> {
+              lock.acquireUninterruptibly();
+              return Next.next("baz");
+            },
+            effectHandler);
+
+    mobiusLoop = builder.startFrom("foo");
+
+    mobiusLoop.dispatchEvent(new TestEvent("bar"));
+    releaseLockAfterDelay(lock, 30);
+    mobiusLoop.dispose();
+  }
+
+  @Test
+  public void updateMayEmitEffectsBeforeReturningFromDispose() throws Exception {
+    // Update should be allowed to emit effects during dispose without anything
+    // bad happening.
+
+    Semaphore lock = new Semaphore(0);
+
+    final MobiusLoop.Builder<String, TestEvent, TestEffect> builder =
+        Mobius.loop(
+            (model, event) -> {
+              lock.acquireUninterruptibly();
+              return Next.dispatch(effects(new SafeEffect("baz")));
+            },
+            effectHandler);
+
+    mobiusLoop = builder.startFrom("foo");
+
+    mobiusLoop.dispatchEvent(new TestEvent("bar"));
+    releaseLockAfterDelay(lock, 30);
+    mobiusLoop.dispose();
+  }
+
```

Comment:
I think the tests below this line are great, but that the ones above should be deleted; the 'non-crashing' bit is already tested by the ones below, so it seems like the ones above are mostly duplication. I know about the 'test one thing' rule, but there are reasons to break every rule occasionally. :)